### PR TITLE
Make queries thread safe

### PR DESCRIPTION
**Pull request by Zoxc**
This makes queries thread safe by removing the query stack and making queries point to their parents. Queries write to the query map when starting and cycles are detected by checking if there's already an entry in the query map. This makes cycle detection O(1) instead of O(n), where `n` is the size of the query stack.

This is mostly corresponds to the method I described [here](https://internals.rust-lang.org/t/parallelizing-rustc-using-rayon/6606).

cc @rust-lang/compiler

r? @michaelwoerister

### CODE REVIEW

File: src/librustc/ty/maps/plumbing.rs
Comment (michaelwoerister) [inline src/librustc/ty/maps/plumbing.rs]: Is there a reason why this line is not just `return Ok(lock);`? I think that would be clearer.

Comment (Zoxc) [reply src/librustc/ty/maps/plumbing.rs]: Yeah, that would be better. The `break` is there because this used to be part of another function.

Comment (Zoxc) [reply src/librustc/ty/maps/plumbing.rs]: I've fixed this.

Comment (michaelwoerister) [reply src/librustc/ty/maps/plumbing.rs]: :+1:

**Comment by pietroalbini** [comment]
Ping from triage @michaelwoerister! The PR author pushed new commits, could you review them?

**Comment by michaelwoerister** [comment]
I haven't forgotten about this, I'm just busy with bug fixing the last few days.

**Comment by nikomatsakis** [comment]
Same here. I have some time scheduled for review later today, though I expect i'm largely going to defer to @michaelwoerister and/or @eddyb here since I don't have a ton of time right now.

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Needs documentation. What is an `ImplicitCtxt`?

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: If we're going to reformat, let's adopt the modern style:

```
pub fn enter_local<F, R>(
    &self,
    arena: &'tcx DroplessArena,
    f: F,
) -> R
where
    F: ...
{
```

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Comment please. (I know... the original code isn't that documented either...c'est la vie)

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Nit: as above, let's adopt the modern style

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Comment: what is a "fully related" context?

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Comment: what does this do? I guess it gets the context from the environment?

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [reply src/librustc/ty/context.rs]: Also, comment =) 

One thing I am wondering about: why does this take an `&ImplicitCtxt` versus creating one locally?

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Comment: what is a "related context"?

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Comment: What is a "context" and how is that different from a "related" context =)

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Pre-existing, but I'm on a roll: let's comment this too...

File: src/librustc/ty/context.rs
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Likewise

File: src/librustc/ty/maps/job.rs
Comment (nikomatsakis) [inline src/librustc/ty/maps/job.rs]: Ideally this would have a module comment pointing into the rustc-guide chapter that describes the design =)

Comment (nikomatsakis) [reply src/librustc/ty/maps/job.rs]: Basically every struct and public function in this module should have some form of doc comment

File: src/librustc/ty/maps/job.rs
Comment (nikomatsakis) [inline src/librustc/ty/maps/job.rs]: Can you please run rustfmt on this file? <3

File: src/librustc/ty/maps/mod.rs
Comment (nikomatsakis) [inline src/librustc/ty/maps/mod.rs]: Hmm. We should probably rename `canonical::QueryResult`, but not in this PR.

File: src/librustc/ty/maps/plumbing.rs
Comment (nikomatsakis) [inline src/librustc/ty/maps/plumbing.rs]: Can we make a custom enum here? This type is sort of incomprehensible. Something with named variants seems better. Like:

```rust
enum TryGetLock {
    NotYetStarted(Guard), // Today: Ok(guard)
    JobCompleted(...), // Today: Err(Ok(...))
    CycleDetected(...), // Today: Err(Err...))
}
```

I guess it means:

- `Ok(guard)` -- no job yet started, lock is still held
- `Err(Ok(...))` -- cache hit, or else we blocked on a job and it completed
- `Err(Err(..))` -- cycle detected

File: src/librustc/ty/maps/plumbing.rs
Comment (nikomatsakis) [inline src/librustc/ty/maps/plumbing.rs]: why is this a macro and not a function?

Comment (nikomatsakis) [reply src/librustc/ty/maps/plumbing.rs]: Oh, because it returns I guess? Hmm. OK. Maybe we can rename it `get_lock_or_return` or something.

File: src/librustc/ty/maps/plumbing.rs
Comment (nikomatsakis) [inline src/librustc/ty/maps/plumbing.rs]: Neat.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Epic PR @Zoxc =)

I did a first read. I'm basically happy for now with the overall setup, which is what I expected. One obvious problem is that the code could use a lot more comments.

This is to some extent a pre-existing problem: the whole query setup is woefully underdocumented as is, but this is as good a time as any to fix that. In general, we should aim to make this stuff maximally readable. 

(Once this has landed, I would like to iterate on some of the details, such as whether to use lazy or eager cycle detection, and the best way to handle blocking jobs -- but that can wait.)

**Comment by Zoxc** [comment]
@Mark-Simulacrum Can I get a perf run?

**Comment by Mark-Simulacrum** [comment]
Perf queued.

**Comment by Mark-Simulacrum** [comment]
http://perf.rust-lang.org/compare.html?start=c19264fa835a1eca86de4fd2e86a87b3919e57cf&end=c83390e645c0ae1d55d6f56636824a76fe084663&stat=instructions%3Au

**Comment by Zoxc** [comment]
I've renamed `StackEntry` to `QueryInfo` since there's no longer a stack. I removed the `track_diagnostics` field of `QueryJob` since it was no longer used. I also added some comments.

**Comment by michaelwoerister** [comment]
Sorry for not getting to this yet, @Zoxc. Either we'll take a look at this together during the work week or I'll do an in-depth review on Tuesday after Easter.

**Comment by michaelwoerister** [comment]
OK, I finally had time to review this. The comments you added since the initial version make things a lot clearer, thanks, @Zoxc! I think we can go ahead an merge this. The basic strategy seems sound to me and a good starting point. 

@nikomatsakis, are you satisfied with the comments added or do you want more? Otherwise, r=me.

**Comment by michaelwoerister** [comment]
I'm going to go ahead and r+ this for now. We can always add more comments later.

@bors r+

**Comment by michaelwoerister** [comment]
:tada:

**Pull request merged**
